src/v2/git_url.rs

```diff
@@ -71,6 +71,44 @@ impl GitUrl {
             }
         }
     }
+
+    /// Returns a new GitUrl which is the same as the
+    /// this one, but without any subdirectory part
+    pub fn without_subdirectory(&self) -> GitUrl {
+        let (repository, branch, _) = self.parse_parts();
+        let branch_str = match branch {
+            Some(branch) => format!("#{}", branch),
+            None => String::new(),
+        };
+        GitUrl { url: format!("{}{}", repository, branch_str) }
+    }
+
+    /// Extract the repository part of the URL
+    pub fn repository(&self) -> String {
+        self.parse_parts().0
+    }
+
+    /// Extract the optional branch part of the git URL
+    pub fn branch(&self) -> Option<String> {
+        self.parse_parts().1
+    }
+
+    /// Extract the optional subdirectory part of the git URL
+    pub fn subdirectory(&self) -> Option<String> {
+        self.parse_parts().2
+    }
+
+    fn parse_parts(&self) -> (String, Option<String>, Option<String>) {
+        lazy_static! {
+            static ref URL_PARSE: Regex = Regex::new(r#"^([^#]+)(#([^:]+)?(:(.+))?)?$"#).unwrap();
```

Comment:
If we were being purely pendantic, it might be nice to change `.unwrap()` to `.expect("could not parse URL_PARSE")`, but that's being a bit pendantic, and much of `compose_yml` is old enough that I hadn't started doing that yet myself. :-)

Either `expect` or `unwrap` is fine here, because this is a true assertion violation.

Also, we might add a couple of  `:?` "no-capture" markers as follows:

```
^([^#]+)(?:#([^:]+)?(?::(.+))?)?$
```

I _think_ those are in the right place. :-) This would also require changing the arguments to `at` below.

Reply:
Aha, I didn't even know about no-capture, that's awesome.